IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,321 @@
 //------------------------------------------------------------------------------
 
 #import "MSIDOAuth2EmbeddedWebviewController.h"
+#import "MSIDTelemetry+Internal.h"
+#import "MSIDTelemetryUIEvent.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSIDAadAuthorityCache.h"
+#import "MSIDError.h"
+#import "MSIDWebOAuth2Response.h"
+#import "MSIDWebviewAuthorization.h"
+#import "MSIDChallengeHandler.h"
+#import "MSIDAuthority.h"
+#import "MSIDWorkPlaceJoinConstants.h"
 
 @implementation MSIDOAuth2EmbeddedWebviewController
+{
+    NSURL *_endUrl;
+    MSIDWebUICompletionHandler _completionHandler;
+    
+    NSLock *_completionLock;
+    NSTimer *_spinnerTimer; // Used for managing the activity spinner
+}
 
+- (id)initWithStartUrl:(NSURL *)startUrl
+                endURL:(NSURL *)endUrl
+               webview:(WKWebView *)webview
+         configuration:(MSIDWebviewConfiguration *)configuration
+               context:(id<MSIDRequestContext>)context
+{
+    self = [super initWithContext:context];
+    
+    if (self)
+    {
+        self.webView = webview;
+        _startURL = startUrl;
+        _endUrl = endUrl;
+        
+        _completionLock = [[NSLock alloc] init];
+        self.complete = NO;
+        
 #if TARGET_OS_IPHONE
-@synthesize  parentViewController;
+        self.parentController = configuration.parentController;
+        self.presentationType = configuration.presentationType;
 #endif
+    }
+    
+    return self;
+}
+
+-(void)dealloc
+{
+    [self.webView setNavigationDelegate:nil];
+    self.webView = nil;
+}
 
 - (BOOL)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
 {
-    return NO;
+    // If we're not on the main thread when trying to kick up the UI then
+    // dispatch over to the main thread.
+    if (![NSThread isMainThread])
+    {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self startWithCompletionHandler:completionHandler];
+        });
+        return YES;
```

Comment:
Since switching to main queue is asynchronized, no return type is allowed for the dispatch block at line 88.
So the solution here is to always return YES and it will always call the completionHandler if it sees any error during the async call.